design/tlog-spilling.md.html

```diff
@@ -0,0 +1,590 @@
+each DiskQueue page to verify data integrity.  If an externally maintained
+checksums exists to verify the returned data, then `CheckHashes::NO` can be
+used to elide the checksumming.
+
+### Popping
+
+As storage servers persist data, they send `pop(tag, version)` requests to the
+transaction log to notify it that it is allowed to discard data for `tag` up
+through `version`.  Once all the tags have been popped from the oldest commit
+in the DiskQueue, the tail of the DiskQueue can be discarded to reclaim space.
+
+
+Each time FoundationDB goes through a recovery, it will recruit a new
+generation of transaction logs.  This new generation of transaction logs will
+often be recruited on the same worker that hosted the previous generation's
+transaction log.  The old generation of transaction logs will only shut down
+once all the data that they have has been fully popped.  This means that there
+can be multiple instances of a transaction log 
+
+*********************************************************
+*                   SharedTLog                          *
+*                                                       *
+* +--------+--------+--------+--------+--------+        *
+* | TLog 1 | TLog 2 | TLog 3 | TLog 4 | TLog 5 |        *
+* +--------+--------+--------+--------+--------+        *
+*   ^ popping         ^spilling         ^committing     * 
+*********************************************************
+
+
+
+
+### Transaction State Store
+
+For FDB to perform a recovery, there is information that it needs to know about
+the database, such as the configuration, worker exclusions, backup status, etc.
+These values are stored into the database in the `\xff` system keyspace.
+However, during a recovery, FDB can't read this data from the storage servers,
+because recovery hasn't completed, so it doesn't know who the storage servers
+are yet.  Thus, a copy of this data is held in-memory on every proxy in the
+*transaction state store*, and durably persisted as a part of commits on the
+transaction logs.  Being durably stored on the transaction logs means the list
+of transaction logs can be fetched from the coordinators, and then used to load
+the rest of the information about the database.
+
+The in-memory storage engine writes an equal amount of mutations and snapshot
+data to a queue, an when a full snapshot of the data has been written, deletes
+the preceeding snapshot and begins writing a new one.  When backing an
+in-memory storage engine with the transaction logs, the
+`LogSystemDiskQueueAdapter` implements writing to a queue as committing
+mutations to the transaction logs with a special tag of `txsTag`, and deleting
+the preceeding snapshot as popping the transaction logs for the tag of `txsTag`
+until the version where the last full snapshot began.
+
+This means that unlike every other commit that is tagged and stored on the
+transaction logs, `txsTag` signifies data that is:
+
+1. Committed to infrequently
```

Comment:
Why is it committed to infrequently? Seems like it should be relatively constant since the in-memory storage engine is constantly writing the last mutations + trying to write out the latest snapshot